src/devices/KeyMatrix/KeyMatrix.cs

```diff
@@ -0,0 +1,211 @@
+        public delegate void PinChangeEventHandler(object sender, KeyMatrixEventArgs pinValueChangedEventArgs);
+
+        private int[] _outputPins;
+        private int[] _inputPins;
+        private GpioController _masterGpioController;
+        private PinValue[] _buttonValues;
+        private bool _pinsOpened;
+
+        /// <summary>
+        /// Initialize key matrix
+        /// </summary>
+        /// <param name="masterController">GPIO controller</param>
+        /// <param name="outputPins">Output pins</param>
+        /// <param name="inputPins">Input pins</param>
+        /// <param name="scanInterval">Scanning interval in milliseconds</param>
+        public KeyMatrix(GpioController masterController, IEnumerable<int> outputPins, IEnumerable<int> inputPins, int scanInterval)
+        {
+            _masterGpioController = masterController ?? throw new ArgumentNullException(nameof(masterController));
+
+            if (outputPins == null)
+            {
+                throw new ArgumentNullException(nameof(outputPins));
+            }
+            if (!outputPins.Any())
+            {
+                throw new ArgumentOutOfRangeException(nameof(outputPins), "The number of outputs is at least 1");
+            }
+
+            if (inputPins == null)
+            {
+                throw new ArgumentNullException(nameof(inputPins));
+            }
+            if (!inputPins.Any())
+            {
+                throw new ArgumentOutOfRangeException(nameof(inputPins), "The number of inputs is at least 1");
+            }
+
+            if (scanInterval <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(scanInterval), "Scanning interval must be positive");
+            }
+
+            _outputPins = outputPins.ToArray();
+            _inputPins = inputPins.ToArray();
+            _buttonValues = new PinValue[_outputPins.Length * _inputPins.Length];
+            _pinsOpened = false;
+
+            ScanInterval = scanInterval;
+
+            OpenPins();
+        }
+
+        /// <summary>
+        /// Start a task to open GPIO pins then start keyboard scanning. Raises KeyMatrixEventArgs.
+        /// </summary>
+        /// <param name="token">A cancellation token that can be used to cancel the work</param>
+        public Task ScanAsync(CancellationToken token)
+        {
+            return Task.Run(() =>
+            {
+                KeyMatrixEventArgs args = Scan(token, false);
+                return args;
+            }, token);
+        }
+
+        /// <summary>
+        /// Start a task to open GPIO pins then start keyboard scanning. End with a KeyMatrixEventArgs raises.
+        /// </summary>
+        /// <param name="token">A cancellation token that can be used to cancel the work</param>
+        public Task<KeyMatrixEventArgs> ReadKeyAsync(CancellationToken token)
+        {
+            return Task.Run(() =>
+            {
+                KeyMatrixEventArgs args = Scan(token, true);
```

Comment:
considering you're auto scanning (at least that's what scanInterval suggests) why do you need to manually scan?